docs/DESIGN.md

```diff
@@ -0,0 +1,122 @@
+
+### CLI
+
+TBD
+
+### Common use cases
+
+#### Firmware update
+
+Per the design of the OpenCyphal's standard network services, the firmware update process is entirely driven by the node being updated (updatee) rather than the node providing the new firmware file (updater). While it is possible to indirectly infer the progress of the update process by observing the offset of the file reads done by the updatee, this solution is fragile because there is ultimately no guarantee that the updatee will read the file sequentially, or even read it in its entirety. Per the OpenCyphal design, the only relevant parameters of a remote node that can be identified robustly are:
+
+- Whether a firmware update is currently in progress or not.
+- The version numbers, CRC, and VCS ID of the firmware that is currently being executed.
+
+The proposed API allows one to commence an update process and wait for its completion as follows:
+
+1. Identify the node that requires a firmware update, and locate a suitable firmware image file on the local machine.
+2. `daemon.get_file_server().add_root(firmware_path)`, where `firmware_path` is the path to the new image.
+3. `daemon.get_node_command_client().begin_software_update(node_id, firmware_name)`, where `firmware_name` is the last component of the `firmware_path`.
+4. Using `daemon.get_monitor().snapshot()`, ensure that the node in question has entered the firmware update mode. Abort if not.
+5. Using `daemon.get_monitor().snapshot()`, wait until the node has left the firmware update mode.
+6. Using `daemon.get_monitor().snapshot()`, ensure that the firmware version numbers match those of the new image.
+
+It is possible to build a convenience method that manages the above steps. Said method will be executed on the client side as opposed the daemon side.
+
+##### Progress monitoring
+
+To enable monitoring the progress of a firmware update process, the following solutions have been considered and rejected:
+
+- Add an additional general-purpose numerical field to `uavcan.node.ExecuteCommand.1` that returns the progress information when an appropriate command (a new standard command) is sent. This is rejected because an RPC-based solution is undesirable.
+
+- Report the progress via `uavcan.node.Heartbeat.1.vendor_specific_status_code`. This is rejected because the VSSC is vendor-specific, so it shouldn't be relied on by the standard.
+
+The plan that is tentatively agreed upon is to define a new standard message with a fixed port-ID for needs of progress reporting. The message will likely be placed in the diagnostics namespace as `uavcan.diagnostic.ProgressReport` with a fixed port-ID of 8183. The `uavcan.node.ExecuteCommand.1` RPC may return a flag indicating if the progress of the freshly launched process will be reported via the new message.
```

Comment:
With fixed port-id approach, for me it looks like there can be potentially only one operation in progress. IMO more flexible approach would be if ExecuteCommand has a new field with desired port id (f.e. with `progressPortId` field name) where server should post `uavcan.diagnostic.ProgressReport`-s for this operation we are executing. If this `progressPortId` is set (>0) then it means that client wants progresses. With this approach...:
- we don't need "_RPC may return a flag indicating if the progress ..._" thing
- RTP client can decide whether progresses are needed or not
- potentially multiple simultaneous commands could be executed, each with their own progress msgs flow

Alternately, especially if a fixed port-ID is desired, we should at least add some unique operation id (int64?) into ExecuteCommand and into corresponding ProgressReport messages, so that it will be possible to distinguish to which operation this particular progress msg is related to. Again, we can use default 0 operation id as indication that client doesn't want progresses for this command.

Reply:
>With this approach...:

I was originally reluctant to agree to this because opening a new publisher on the updatee is a highly fallible operation that is better done at the initialization stage, but then I concluded that the advantages probably outweigh the disadvantages. You have already enumerated the advantages of this approach correctly. I might add that the fixed port-ID solution is too reminiscent of DroneCAN in its rigidity, and if one wants DroneCAN, one knows where to find it.

In the most basic case, each node sending ExecuteCommand requests would have a subscriber for the progress reports configured via a register; the same subject-ID will be sent to the clients when progress reports are desired.

In this case we don't need a new message type as we can just leverage `uavcan.primitive.scalar.Natural8.1`.

Reply:
@lydia-at-amazon That would work, but: we considered a similar solution where we add a progress field to the response of `uavcan.protocol.ExecuteCommand`. Scott pointed out that an RPC-based solution is considered harmful, so we focused on pub/sub based alternatives. If RPC is acceptable, then I think your current register-based approach is probably the most obvious choice.

Reply:
For maintenance operations, such as software update, we're already using lots of RPC services, so I think it should be ok to use the RPC-based solution.